"IllegalArgumentException: LocalVariablesSorter only accepts expanded frames" when logging in PanacheRepository

### Describe the bug

Using Quarkus logging in a PanacheRepository crashes the build. I'm using Kotlin, haven't tried in with Java.

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

1. Start a new (Kotlin) Quarkus project, for the reproducer I did `quarkus create app org.acme:rest-kotlin-quickstart --extension='kotlin,resteasy-reactive-jackson'`
2. Add these extensions `quarkus-jdbc-postgresql` and `quarkus-hibernate-orm-panache-kotlin`
3. Create a class extending PanacheRepository
4. Create a data class with @Entity annotation for the repository type since PanacheRepository is generic
5. Create a function that tries to log using e.g. `Log.error("Test Quarkus logging")` or `Log.info("Test Quarkus logging")`
6. Try to build and observe the failure.

See attached reproducer.

[rest-kotlin-quickstart.zip](https://github.com/quarkusio/quarkus/files/10345223/rest-kotlin-quickstart.zip)


### Output of `uname -a` or `ver`

Linux pop-os 5.19.16-76051916-generic #202210150742\~1666053244\~22.04\~cf07008 SMP PREEMPT_DYNAMIC Tue O x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk 17.0.4.1 2022-08-12 OpenJDK Runtime Environment Temurin-17.0.4.1+1 (build 17.0.4.1+1) OpenJDK 64-Bit Server VM Temurin-17.0.4.1+1 (build 17.0.4.1+1, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

Reproducer: 2.15.1.Final, project where I noticed the bug: 2.14.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Reproducer: Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63), project where I noticed the bug: Gradle 7.5.1

### Additional information

Stacktrace: 
```
2023-01-04 16:06:31,479 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (vert.x-worker-thread-0) Failed to start quarkus [Error Occurred After Shutdown]: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.steps.ClassTransformingBuildStep#handleClassTransformation threw an exception: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:918)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep.handleClassTransformation(ClassTransformingBuildStep.java:258)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
        ... 6 more
Caused by: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at org.objectweb.asm.commons.LocalVariablesSorter.visitFrame(LocalVariablesSorter.java:188)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.emitNullCheck(KotlinPanacheClassOperationGenerationVisitor.java:300)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.nullCheckReturn(KotlinPanacheClassOperationGenerationVisitor.java:513)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.invokeOperation(KotlinPanacheClassOperationGenerationVisitor.java:452)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.generate(KotlinPanacheClassOperationGenerationVisitor.java:341)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.visitEnd(KotlinPanacheClassOperationGenerationVisitor.java:557)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:748)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep.transformClass(ClassTransformingBuildStep.java:352)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:223)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:210)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        ... 5 more

        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:335)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:265)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:60)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:220)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.restartCallback(IsolatedDevModeMain.java:203)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:537)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:437)
        at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:152)
        at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:139)
        at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
        at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.steps.ClassTransformingBuildStep#handleClassTransformation threw an exception: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:918)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep.handleClassTransformation(ClassTransformingBuildStep.java:258)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
        ... 6 more
Caused by: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at org.objectweb.asm.commons.LocalVariablesSorter.visitFrame(LocalVariablesSorter.java:188)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.emitNullCheck(KotlinPanacheClassOperationGenerationVisitor.java:300)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.nullCheckReturn(KotlinPanacheClassOperationGenerationVisitor.java:513)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.invokeOperation(KotlinPanacheClassOperationGenerationVisitor.java:452)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.generate(KotlinPanacheClassOperationGenerationVisitor.java:341)                                                                          sitor.visitEnd(KotlinPanacheClassOperationGenerationVisitor.java:557)at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVi
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:748)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep.transformClass(ClassTransformingBuildStep.java:352)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:223)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:210)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        ... 5 more

        at io.quarkus.builder.Execution.run(Execution.java:123)
        at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:331)
        ... 18 more
Caused by: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:918)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep.handleClassTransformation(ClassTransformingBuildStep.java:258)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
        ... 6 more
Caused by: java.lang.IllegalArgumentException: LocalVariablesSorter only accepts expanded frames (see ClassReader.EXPAND_FRAMES)
        at org.objectweb.asm.commons.LocalVariablesSorter.visitFrame(LocalVariablesSorter.java:188)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.emitNullCheck(KotlinPanacheClassOperationGenerationVisitor.java:300)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.nullCheckReturn(KotlinPanacheClassOperationGenerationVisitor.java:513)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.invokeOperation(KotlinPanacheClassOperationGenerationVisitor.java:452)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.generate(KotlinPanacheClassOperationGenerationVisitor.java:341)
        at io.quarkus.panache.common.deployment.visitors.KotlinPanacheClassOperationGenerationVisitor.visitEnd(KotlinPanacheClassOperationGenerationVisitor.java:557)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:748)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:424)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep.transformClass(ClassTransformingBuildStep.java:352)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:223)
        at io.quarkus.deployment.steps.ClassTransformingBuildStep$2.call(ClassTransformingBuildStep.java:210)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        ... 5 more
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/BytecodeTransformerBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/BytecodeTransformerBuildItem.java
@@ -44,6 +44,8 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
 
     final boolean continueOnFailure;
 
+    final int priority;
+
     public BytecodeTransformerBuildItem(String classToTransform,
             BiFunction<String, ClassVisitor, ClassVisitor> visitorFunction) {
         this(classToTransform, visitorFunction, null);
@@ -81,6 +83,7 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
         this.inputTransformer = null;
         this.classReaderOptions = 0;
         this.continueOnFailure = false;
+        this.priority = 0;
     }
 
     public BytecodeTransformerBuildItem(Builder builder) {
@@ -92,6 +95,7 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
         this.inputTransformer = builder.inputTransformer;
         this.classReaderOptions = builder.classReaderOptions;
         this.continueOnFailure = builder.continueOnFailure;
+        this.priority = builder.priority;
         if (visitorFunction == null && inputTransformer == null) {
             throw new IllegalArgumentException("One of either visitorFunction or inputTransformer must be set");
         }
@@ -129,6 +133,28 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
         return continueOnFailure;
     }
 
+    /**
+     * Bytecode transformers are applied in ascending priority order. That is, lower priority
+     * value means the transformer is applied sooner, and higher priority value means
+     * the transformer is applied later.
+     * <p>
+     * This applies directly to {@link #inputTransformer} functions: an input transformer
+     * function with lower priority is applied first and its result is passed to the transformer
+     * function with higher priority.
+     * <p>
+     * It is a bit counter-intuitive when it comes to the {@link #visitorFunction}. The visitor
+     * function doesn't directly transform bytecode; instead, it constructs an ASM
+     * {@link ClassVisitor} from an earlier class visitor. This means that the priority order
+     * is effectively turned around: the later a bytecode transformer is called,
+     * the <em>higher</em> in the class visitor chain it ends up, and the <em>sooner</em>
+     * is the visitor eventually called.
+     * <p>
+     * The priority value defaults to {@code 0}.
+     */
+    public int getPriority() {
+        return priority;
+    }
+
     public static class Builder {
         public BiFunction<String, byte[], byte[]> inputTransformer;
         public boolean continueOnFailure;
@@ -138,6 +164,7 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
         private boolean eager = false;
         private boolean cacheable = false;
         private int classReaderOptions = 0;
+        private int priority = 0;
 
         public Builder setContinueOnFailure(boolean continueOnFailure) {
             this.continueOnFailure = continueOnFailure;
@@ -176,6 +203,11 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
 
         public Builder setClassReaderOptions(int classReaderOptions) {
             this.classReaderOptions = classReaderOptions;
+            return this;
+        }
+
+        public Builder setPriority(int priority) {
+            this.priority = priority;
             return this;
         }
 

--- a/core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingWithPanacheProcessor.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingWithPanacheProcessor.java
@@ -36,6 +36,7 @@ public class LoggingWithPanacheProcessor {
                     .setClassToTransform(className)
                     .setVisitorFunction((ignored, visitor) -> new AddLoggerFieldAndRewriteInvocations(visitor, className))
                     .setClassReaderOptions(ClassReader.EXPAND_FRAMES)
+                    .setPriority(1000)
                     .build());
         }
     }

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
@@ -9,6 +9,7 @@ import java.nio.file.StandardOpenOption;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -128,10 +129,14 @@ public class ClassTransformingBuildStep {
                         .filter(a -> !a.isContinueOnFailure())
                         .findAny().isEmpty();
                 List<BiFunction<String, ClassVisitor, ClassVisitor>> visitors = classTransformers.stream()
-                        .map(BytecodeTransformerBuildItem::getVisitorFunction).filter(Objects::nonNull)
+                        .sorted(Comparator.comparingInt(BytecodeTransformerBuildItem::getPriority))
+                        .map(BytecodeTransformerBuildItem::getVisitorFunction)
+                        .filter(Objects::nonNull)
                         .collect(Collectors.toList());
                 List<BiFunction<String, byte[], byte[]>> preVisitFunctions = classTransformers.stream()
-                        .map(BytecodeTransformerBuildItem::getInputTransformer).filter(Objects::nonNull)
+                        .sorted(Comparator.comparingInt(BytecodeTransformerBuildItem::getPriority))
+                        .map(BytecodeTransformerBuildItem::getInputTransformer)
+                        .filter(Objects::nonNull)
                         .collect(Collectors.toList());
                 ClassLoader old = Thread.currentThread().getContextClassLoader();
                 try {
@@ -201,10 +206,14 @@ public class ClassTransformingBuildStep {
                         .filter(a -> !a.isContinueOnFailure())
                         .findAny().isEmpty();
                 List<BiFunction<String, ClassVisitor, ClassVisitor>> visitors = entry.getValue().stream()
-                        .map(BytecodeTransformerBuildItem::getVisitorFunction).filter(Objects::nonNull)
+                        .sorted(Comparator.comparingInt(BytecodeTransformerBuildItem::getPriority))
+                        .map(BytecodeTransformerBuildItem::getVisitorFunction)
+                        .filter(Objects::nonNull)
                         .collect(Collectors.toList());
                 List<BiFunction<String, byte[], byte[]>> preVisitFunctions = entry.getValue().stream()
-                        .map(BytecodeTransformerBuildItem::getInputTransformer).filter(Objects::nonNull)
+                        .sorted(Comparator.comparingInt(BytecodeTransformerBuildItem::getPriority))
+                        .map(BytecodeTransformerBuildItem::getInputTransformer)
+                        .filter(Objects::nonNull)
                         .collect(Collectors.toList());
                 transformedToArchive.put(classFileName, jar);
                 transformed.add(buildExecutor.submit(new Callable<TransformedClassesBuildItem.TransformedClass>() {

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/KotlinPanacheClassOperationGenerationVisitor.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/KotlinPanacheClassOperationGenerationVisitor.java
@@ -65,6 +65,7 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
     public static final ByteCodeType OBJECT = new ByteCodeType(Object.class);
     protected static final ByteCodeType CLASS = new ByteCodeType(Class.class);
     private static final String CTOR_METHOD_NAME = "<init>";
+    private static final String CLINIT_METHOD_NAME = "<clinit>";
     protected final Function<String, String> argMapper;
     protected final ClassInfo classInfo;
     protected final ByteCodeType entityUpperBound;
@@ -541,8 +542,9 @@ public class KotlinPanacheClassOperationGenerationVisitor extends ClassVisitor {
         } else if (name.contains("$")) {
             //some agents such as jacoco add new methods, they generally have $ in the name
             return super.visitMethod(access, name, descriptor, signature, exceptions);
-        } else if (name.equals(CTOR_METHOD_NAME)) {
+        } else if (name.equals(CTOR_METHOD_NAME) || name.equals(CLINIT_METHOD_NAME)) {
             //Arc can add no-args constructors to support intercepted beans
+            // Logging with Panache can add a class initializer
             return super.visitMethod(access, name, descriptor, signature, exceptions);
         }
         return null;

--- a/integration-tests/logging-panache-kotlin/pom.xml
+++ b/integration-tests/logging-panache-kotlin/pom.xml
@@ -1,0 +1,131 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <artifactId>quarkus-integration-tests-parent</artifactId>
+        <groupId>io.quarkus</groupId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-integration-test-logging-panache-kotlin</artifactId>
+
+    <name>Quarkus - Integration Tests - Logging with Panache in Kotlin</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm-panache-kotlin</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jdbc-h2</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc-deployment</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5-internal</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <!-- Minimal test dependencies to *-deployment artifacts for consistent build order -->
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm-panache-kotlin-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jdbc-h2-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <defaultGoal>install</defaultGoal>
+        <sourceDirectory>src/main/kotlin</sourceDirectory>
+        <testSourceDirectory>src/test/kotlin</testSourceDirectory>
+        <plugins>
+            <plugin>
+                <groupId>org.jetbrains.kotlin</groupId>
+                <artifactId>kotlin-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>compile</id>
+                        <phase>compile</phase>
+                        <goals>
+                            <goal>compile</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>test-compile</id>
+                        <phase>test-compile</phase>
+                        <goals>
+                            <goal>test-compile</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.jetbrains.kotlin</groupId>
+                        <artifactId>kotlin-maven-allopen</artifactId>
+                        <version>${kotlin.version}</version>
+                    </dependency>
+                </dependencies>
+                <configuration>
+                    <compilerPlugins>
+                        <plugin>all-open</plugin>
+                    </compilerPlugins>
+                    <pluginOptions>
+                        <option>all-open:annotation=javax.ws.rs.Path</option>
+                    </pluginOptions>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>build</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>

--- a/integration-tests/logging-panache-kotlin/src/main/kotlin/io/quarkus/logging/LoggingEntity.kt
+++ b/integration-tests/logging-panache-kotlin/src/main/kotlin/io/quarkus/logging/LoggingEntity.kt
@@ -1,0 +1,11 @@
+package io.quarkus.logging
+
+import io.quarkus.hibernate.orm.panache.kotlin.PanacheEntity
+import javax.persistence.Entity
+
+@Entity
+open class LoggingEntity : PanacheEntity() {
+    fun test() {
+        Log.info("Hello from entity")
+    }
+}

--- a/integration-tests/logging-panache-kotlin/src/main/kotlin/io/quarkus/logging/LoggingRepository.kt
+++ b/integration-tests/logging-panache-kotlin/src/main/kotlin/io/quarkus/logging/LoggingRepository.kt
@@ -1,0 +1,11 @@
+package io.quarkus.logging
+
+import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepository
+import javax.enterprise.context.ApplicationScoped
+
+@ApplicationScoped
+open class LoggingRepository : PanacheRepository<LoggingEntity> {
+    fun test() {
+        Log.info("Hello from repository")
+    }
+}

--- a/integration-tests/logging-panache-kotlin/src/test/kotlin/io/quarkus/logging/LoggingWithPanacheKotlinTest.java
+++ b/integration-tests/logging-panache-kotlin/src/test/kotlin/io/quarkus/logging/LoggingWithPanacheKotlinTest.java
@@ -1,0 +1,39 @@
+package io.quarkus.logging;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+import java.util.logging.Formatter;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.jboss.logmanager.formatters.PatternFormatter;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class LoggingWithPanacheKotlinTest {
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar.addClasses(LoggingEntity.class, LoggingRepository.class))
+            .setLogRecordPredicate(record -> record.getLoggerName().startsWith("io.quarkus.logging.Logging"))
+            .assertLogRecords(records -> {
+                Formatter formatter = new PatternFormatter("[%p] %m");
+                List<String> lines = records.stream().map(formatter::format).map(String::trim).collect(Collectors.toList());
+
+                assertThat(lines).containsExactly(
+                        "[INFO] Hello from repository",
+                        "[INFO] Hello from entity");
+            });
+
+    @Inject
+    LoggingRepository repository;
+
+    @Test
+    public void test() {
+        repository.test();
+        new LoggingEntity().test();
+    }
+}

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -323,6 +323,7 @@
                 <module>logging-min-level-unset</module>
                 <module>logging-min-level-set</module>
                 <module>logging-panache</module>
+                <module>logging-panache-kotlin</module>
                 <module>locales</module>
                 <module>redis-devservices</module>
                 <!-- gRPC tests -->

